src/testdrive/src/action/postgres/execute.rs

```diff
@@ -33,26 +33,32 @@ impl Action for ExecuteAction {
         Ok(())
     }
 
-    async fn redo(&self, _: &mut State) -> Result<(), String> {
-        let (client, conn) = tokio_postgres::connect(&self.connection, NoTls)
-            .await
-            .map_err(|e| format!("connecting to postgres: {}", e))?;
-        println!(
-            "Executing queries against PostgreSQL server at {}...",
-            self.connection
-        );
-        let conn_handle = tokio::spawn(conn);
+    async fn redo(&self, state: &mut State) -> Result<(), String> {
+        let client = if self.connection.starts_with("postgres://") {
+            postgres_client(&self.connection).await?
+        } else {
+            state
+                .postgres_clients
+                .remove(&self.connection)
```

Comment:
I spend a futile two hours of my life trying to get Rust to alow me to borrow the hashmap entry for temporary use ... the only way I could make it work is to ```remove``` it from the hashmap and then insert it again after executing the queries :-(

Reply:
@philip-stoev I think the problem might not have been related to borrowing per se, but to the fact that the if branches return different types? `HashMap::get(String)` returns `&Client` (borrowed) whereas `postgres_client(String)` returns `Client` (owned).

One way to illustrate it is as follows (**not saying that you should change it like this**):


```rust
if self.connection.starts_with("postgres://") {
    let client = postgres_client(&self.connection).await?;
    for query in &self.queries {
        println!(">> {}", query);
        client
            .batch_execute(query)
            .await
            .map_err(|e| format!("executing postgres query: {}", e))?;
    }
    drop(client);
} else {
    let client = state
        .postgres_clients
        .get(&self.connection)
        .ok_or(format!("connection '{}' not found", &self.connection))?;
    for query in &self.queries {
        println!(">> {}", query);
        client
            .batch_execute(query)
            .await
            .map_err(|e| format!("executing postgres query: {}", e))?;
    }
}
```

Reply:
Yes, I was also considering that but it involves duplicating code, which I wanted to avoid. Admittedly, my kludge is worse.

I was hoping for some rust-ism that will allow me to consolidate the two code paths without using ```remove``` ...

Reply:
I think you're probably looking for something like this?

```rust
    async fn redo(&self, state: &mut State) -> Result<(), String> {
        let client;
        let client = if self.connection.starts_with("postgres://") {
            client = postgres_client(&self.connection).await?;
            &client
        } else {
            state
                .postgres_clients
                .get(&self.connection)
                .ok_or(format!("connection '{}' not found", &self.connection))?
        };

        for query in &self.queries {
            println!(">> {}", query);
            client
                .batch_execute(query)
                .await
                .map_err(|e| format!("executing postgres query: {}", e))?;
        }

        Ok(())
    }
```

Reply:
Thank you, this is the construct I was looking for !